src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -852,6 +852,10 @@ internal ReadResult GetReadAsyncResult()
             finally
             {
                 cancellationTokenRegistration.Dispose();
+            }
+
+            if (result.IsCanceled)
```

Comment:
ReadResult.IsCanceled should have nothing to do with the CancellationTokens or OperationCanceledExceptions. The race we're trying to fix should remain broken in the event someone also happened to call CancelPendingRead().

PipeAwaitable.CancellationTokenFired() should transition the awaitable to a different AwaitableState than PipeAwaitable.Cancel() does. I think we're going to need add a new AwaitableState. Maybe call it CancellationTokenFired. We should verify we're in this new AwaitableState before throwing any OperationCanceledExceptions.

Reply:
I don't think it's worth it. There's no race for the cancel pending read case per se. ObserveCancellation will read a snapshot and will fix the state. The issue with token is that it's mutated we read the result

Reply:
The extra state would guard against throwing, if both the token *and* CancelPendingRead fire at the same time. I don't know if the difference needs to be observable.

Reply:
Talked with Stephen, and this PR is good as is, resolves the race where the Pipe becomes broken. There is still a break in contract in the case where a `CancelPendingRead` fired and a token was also canceled, instead of receiving a canceled `ReadResult` for the `CancelPendingRead` (which is what the contract implies), you'll receive the exception.

Not worth fixing in this PR though.

Reply:
The issue isn't just that you observe an OperationCanceledException instead of ReadResult.IsCanceled when the CancellationToken fires right as you retrieve the result from the ValueTask. That, by itself, is not even a bug in my opinion considering the CancellationToken did in fact fire. The fact that no read afterwards will return a canceled ReadResult either is the real bug.

It's common for code to be written with the expectation that it's guaranteed to get a canceled ReadResult after calling CancelPendingRead() (assuming the pipe isn't already completed). What's less common is to continue reading after an OperationCanceledException. But if that never happened, we wouldn't be fixing this issue in the first place.